server/src/main/java/io/crate/planner/optimizer/symbol/rule/SwapCastsInComparisonOperators.java

```diff
@@ -60,10 +60,16 @@ private static boolean isSwappable(Function cmpOp) {
             Reference ref = (Reference) cast.arguments().getFirst();
             DataType<?> refInnerType = ArrayType.unnest(ref.valueType());
             DataType<?> literalInnerType = ArrayType.unnest(literal.valueType());
+            var value = literal.value();
             if (!DataTypes.isNumeric(literalInnerType) || !DataTypes.isNumeric(refInnerType)) {
-                return true;
+                try {
+                    refInnerType.implicitCast(value);
+                    return true;
+                } catch (IllegalArgumentException | ClassCastException e) {
+                    return false;
+                }
```

Comment:
How about
```
return literal.valueType().isConvertableTo(ref.valueType(), false);
```
?

Reply:
Looks better, thanks. Applied with an adjustment - used inner types of the Literal and Ref (we do unnest above)